overlord/configstate/config/transaction.go

```diff
@@ -66,6 +66,28 @@ func (t *Transaction) State() *state.State {
 	return t.state
 }
 
+func changes(cfgStr string, cfg map[string]interface{}) []string {
+	var out []string
+	for k := range cfg {
+		switch subCfg := cfg[k].(type) {
+		case map[string]interface{}:
+			out = append(out, changes(cfgStr+"."+k, subCfg)...)
```

Comment:
I've had a quick look at the configstate package to remind myself of how this works, and I _think_ the logic in this function is actually doing the right thing, in the sense that a complete replacement ends up as a RawMessage and a partial replacement ends up as a map. That said, this logic is subtle.. I've tried to add a lot of tests back when first writing the initial bits so that we could remain sure that it'd continue to work. Looking at the tests in this package, it feels a bit lightweight. Can we have some more tests checking these non-obvious conditions report what was changed correctly (three-level down nested partial replacements vs. full replacements, things preserved not being flagged as changes, etc)?

We should eventually make that logic be available through the API for configuration hooks. It's very nice to be able to tell what changed exactly. Not for this PR of course.

Reply:
Indeed, I will improve the tests today!

Reply:
I did improve the tests at the same day :) However the tests showed that nested maps where not handled by `changes()` and fixing this took a bit (mostly because of other more important work). This is fixed now and the tests should cover this quite well now (the existing testcode there is really nice!). But please have a careful and critical look if I missed anything!